### DISCUSSION TITLE
Adding generic components to an entity

### DISCUSSION THREAD

**Discussion by Hellwoodoo** [Quillraven/Fleks#135] [Q&A] [answered]
Just wondering why 
```   
    fun createEntity(vararg component: ComponentType<*>) : Entity { 
          var entity : Entity = sceneWorld.entity {
              it += component
         }
    }
```
does not work with plusAssign, and if there is an alterative to doing this or if I am doing something wrong.
Thanks!

**Comment by Quillraven** [comment, marked as answer]
Hi,

the reason is that you are not adding a component **type** to an entity. You want to add a component itself (an instance of a component). That's why Fleks internally has a function like this (note the component argument instead of component type):

```kotlin
 inline operator fun <reified T : Component<T>> Entity.plusAssign(component: T) {
        val compType: ComponentType<T> = component.type()
        compMasks[this.id].set(compType.id)
        val holder: ComponentsHolder<T> = componentService.holder(compType)
        holder[this] = component
    }
```

There is also an unsafe version that takes a list of components that might fit your use-case better:

```kotlin
 operator fun Entity.plusAssign(components: List<Component<*>>) {
        components.forEach { cmp ->
            val compType = cmp.type()
            compMasks[this.id].set(compType.id)
            val holder = componentService.wildcardHolder(compType)
            holder.setWildcard(this, cmp)
        }
    }
```

**Reply by Hellwoodoo** [reply]
Thanks for that 👍 , managed to get it all sorted out

**Reply by Hellwoodoo** [reply]
@Quillraven In addition, also have a question regarding entities since i'm relatively new to ECS, but should i be referencing internally via their id, or via the entity itself

**Reply by Quillraven** [reply]
Since a few versions you can safely reference entities in Fleks. An entity is nothing else than an ID plus an additional Generation counter (=version).

Imagine you have an entity with ID 1 and you reference it somewhere. Now somewhere else in your code you are removing this entity. 
This will increase the version of this entity by 1 and if you check "entity in world" then it will return false.

If you create a new entity and it gets ID 1 again then its version will be higher than the previous one.

The only thing to keep in mind when referencing entities is that you need to make a check if it is still in your world before doing anything with it. A simple if like before with "entity in world" is sufficient and afterwards execute your code.


Before, without the version, it was problematic to reference entities because of the recycling ID nature of ECS you could have an entity with the same ID but it is a completely different one from the one you initially referenced. Now with the additional generation (=version) information it is a lot better imo.

**Reply by Quillraven** [reply]
Here is a link of the wiki that explains it better. At the end of the page:

https://github.com/Quillraven/Fleks/wiki/Entity
